server/src/main/java/org/elasticsearch/cluster/routing/allocation/ModelHost.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.RoutingNode;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * A model host hold all the node on it.
+ *
+ */
+public final class ModelHost implements Iterable<RoutingNode> {
+    private final List<RoutingNode> nodes = new LinkedList<>();
+
+    public void addNode(RoutingNode node) {
+        nodes.add(node);
+    }
```

Comment:
Can we turn this into an immutable object instead, i.e., build up the list of nodes in `Hosts` constructor and then create this `Host` object in one go with the list?

Reply:
If the `Host` object is create after the build of `nodes` list, we need something like `hostId` to identify a host and temporary data structures like `Map<String, String> hostName2HostIdMap`, `Map<String, String> hostAddress2HostIdMap`, `Map<String, List<RoutingNode> hostId2NodesMap` in `Hosts` constructor. Finally, the host object can be constructed through these temporary data structures. However, it will make the code less clear than it is now. 
How about call a `freeze` function aster one go to make the list immutable.
```
public void freeze() {
    nodes = List.copyOf(nodes);
}
```

Reply:
Maybe I can add a `HostBuilder` class for collect nodes and turn it into a immutable list in `build()`